quickwit/quickwit-indexing/src/source/pulsar_source.rs

```diff
@@ -0,0 +1,1083 @@
+use quickwit_metastore::checkpoint::{
+    PartitionId, Position, SourceCheckpoint, SourceCheckpointDelta,
+};
+use serde_json::{json, Value as JsonValue};
+use tokio::sync::{mpsc, watch};
+use tokio::task::LocalSet;
+use tokio::time;
+use tokio::time::timeout;
+use tracing::{debug, error, info, warn};
+
+use crate::actors::DocProcessor;
+use crate::models::RawDocBatch;
+use crate::source::{
+    Source, SourceActor, SourceContext, SourceExecutionContext, TypedSourceFactory,
+};
+
+/// Number of bytes after which we cut a new batch.
+///
+/// We try to emit chewable batches for the indexer.
+/// One batch = one message to the indexer actor.
+///
+/// If batches are too large:
+/// - we might not be able to observe the state of the indexer for 5 seconds.
+/// - we will be needlessly occupying resident memory in the mailbox.
+/// - we will not have a precise control of the timeout before commit.
+///
+/// 5MB seems like a good one size fits all value.
+const BATCH_NUM_BYTES_LIMIT: u64 = 5_000_000;
+
+/// The duration that the pulsar consumer waits for a new message
+/// to be dispatched.
+const MESSAGE_WAIT_TIMEOUT: Duration = Duration::from_secs(1);
+
+type ConsumerMessage = Result<Message<PulsarMessage>, pulsar::Error>;
+
+pub struct PulsarSourceFactory;
+
+#[async_trait]
+impl TypedSourceFactory for PulsarSourceFactory {
+    type Source = PulsarSource;
+    type Params = PulsarSourceParams;
+
+    async fn typed_create_source(
+        ctx: Arc<SourceExecutionContext>,
+        params: PulsarSourceParams,
+        checkpoint: SourceCheckpoint,
+    ) -> anyhow::Result<Self::Source> {
+        PulsarSource::try_new(ctx, params, checkpoint).await
+    }
+}
+
+#[derive(Default)]
+pub struct PulsarSourceState {
+    /// Number of bytes processed by the source.
+    pub num_bytes_processed: u64,
+    /// Number of messages processed by the source (including invalid messages).
+    pub num_messages_processed: u64,
+    /// Number of invalid messages, i.e., that were empty or could not be parsed.
+    pub num_invalid_messages: u64,
+}
+
+pub struct PulsarSource {
+    ctx: Arc<SourceExecutionContext>,
+    pulsar: PulsarConsumer,
+    params: PulsarSourceParams,
+    subscription_name: String,
+    previous_positions: BTreeMap<PartitionId, Position>,
+    state: PulsarSourceState,
+}
+
+impl PulsarSource {
+    pub async fn try_new(
+        ctx: Arc<SourceExecutionContext>,
+        params: PulsarSourceParams,
+        checkpoint: SourceCheckpoint,
+    ) -> anyhow::Result<Self> {
+        let subscription_name = subscription_name(&ctx.index_id, &ctx.source_config.source_id);
+        info!(
+            index_id=%ctx.index_id,
+            source_id=%ctx.source_config.source_id,
+            topics=?params.topics,
+            subscription_name=%subscription_name,
+            "Create Pulsar source."
+        );
+
+        let pulsar = connect_pulsar(&params).await?;
+
+        let mut previous_positions = BTreeMap::new();
```

Comment:
Could be nice to know how partition ID are derived from the topic name. Otherwise, we could think that partitions IDs could clash between topics.

Reply:
`"{topic}-partition-{ID}"`